### DISCUSSION TITLE
Making WezTerm-NeoVim default text editor on Mac?

### DISCUSSION THREAD

**Discussion by twisteria** [wezterm/wezterm#2316] [Q&A] [answered]
I've been searching for the way to set terminal NeoVim (not GUI app) as a default app in Mac (so that when I double click on a file with code, it opens with terminal NeoVim).

I found StackOverflow discussion to enable this on iTerm (with AppleScript): [url](https://stackoverflow.com/questions/66298660/how-do-i-make-neovim-my-default-text-code-editor)

However, I want to, of course, do the same with WezTerm.
Changing `iTerm` in the script to `WezTerm` did not work.
Since I have absolutely no knowledge of AppleScript, I have no clue going forward.

Could any professionals code it for me and other WezTerm-NeoVim fans?
Or, if this sounds like a good additional function topic for WezTerm -- like Kitty/s case (also mentioned in the above link)?

**Comment by MuhammedZakir** [comment, marked as answer]
WezTerm supports opening an arbitrary command at startup via `start` cli option. E.g. `wezterm-gui start -- nvim somefile`.

To do what you want, you can use the example regarding Kitty, but replace the shell script with `wezterm-gui start -- nvim "$@"`. If you want to use AppleScript, something like below should work.

Note: 2 hiphens(`--`) starts a comment in AppleScript, I did a backslash-escape, but I don't know if that's correct. Paste below script in Script Editor, and click the hammer icon (iirc), and it will syntax highlight the code, which should show whether escaping has solved it.

Edit: It looks like you don't need to escape hiphen when it's inside string? At least, that's what Github's syntax highlighting tells. I have removed the backslash before hiphen. Regardless, it's better if you check this using Script Editor.

```applescript
on run {input, parameters}
  set cmd to "nvim"
  if input is not {} then
    set filePath to POSIX path of input
    set cmd to "nvim \"" & filePath & "\""
  end if

  do shell script "wezterm-gui start -- " & cmd
end run

**Reply by twisteria** [reply]
Thank you so much!

I got it working as expected!
Though it did not work right away - I got "command not found" error.
seems those (3rd party?) commands need full path being addressed as below.. 
```
on run {input, parameters}
	set cmd to "nvim"
	if input is not {} then
		set filePath to POSIX path of input
		set cmd to "/usr/local/bin/nvim \"" & filePath & "\""
	end if
	
	do shell script "/usr/local/bin/wezterm-gui start -- " & cmd
end run
```
Thank you again for your very switf support!!

**Reply by MuhammedZakir** [reply]
It looks like I am out of touch with macOS. :-)

I suggest you to modify the `PATH` instead. This way, you can use locally installed programs from `nvim`.
```applescript
do shell script "env PATH=/usr/local/bin:\"$PATH\" wezterm-gui start -- " & cmd
```

**Reply by twisteria** [reply]
Thanks for the followup!
(honestly, I do not understand the difference, but confirmed both works fine ;)

One additional thing if I may ask...
I found the script only works for one file (one double click).
for the first file, it opens w/o problems, but leaving the window open, double clicking the second file does nothing (to be precise, it shows me macOS's file-opening effect, but it does not open).
When I close the first file, then the second file opens.
Is this a restriction on AppleScript/Automator side, maybe?
or something can be done in the script so to allow multiple files open in multiple windows/tabs?

**Reply by MuhammedZakir** [reply]
> Thanks for the followup!

You're welcome!

> (honestly, I do not understand the difference, but confirmed both works fine ;)

Simply put: by adding `/usr/local/bin` to `$PATH`, you an run any executable in that directory using their names instead of using full path. That is, you can run `nvim`, `wezterm-gui`, etc. instead of having to use full path - `/usr/local/bin/nvim`, `/usr/local/bin/wezterm-gui`, etc.

For more details, search for "PATH environenment variable". 
- https://www.ibm.com/docs/en/aix/7.2?topic=accounts-path-environment-variable
- https://askubuntu.com/questions/141718/what-is-the-path-environment-variable-and-how-do-i-add-to-it

> One additional thing if I may ask... I found the script only works for one file (one double click). for the first file, it opens w/o problems, but leaving the window open, double clicking the second file does nothing (to be precise, it shows me macOS's file-opening effect, but it does not open). When I close the first file, then the second file opens. Is this a restriction on AppleScript/Automator side, maybe? or something can be done in the script so to allow multiple files open in multiple windows/tabs?

Do you get this behavior even after adding `/usr/loca/bin` to the `PATH`? Does adding `--always-new-process` flag work? -- `wezterm-gui start --always-new-process -- `.

**Reply by twisteria** [reply]
Wow, sir!
I didn't expect to actually learn things, besides getting a solution to my issue.
You are incredible! Thank you!

(even though I am aware that I am off the topic that this discussion board should cover, I cannot help rephrasing my understanding)
So, I was thinking I was specifying an absolute route(path) to the command with ```usr/local/bin/nvim`` but this was wrong; if  there exists a different command with same folder structure and command name under one of the paths, there is a chance the system executes wrong commands. How we specify commands path in the correct way is the above feedback from you.

Getting back to my problem, I just checked ```--always-new-process``` flag, but I couldn't find any difference. how I specify the path to the command did not change the result either...

**Reply by MuhammedZakir** [reply]
Just to make sure: did you change the `do shell ...` line to this when you tested it last time?
```applescript
do shell script "env PATH=/usr/local/bin:\"$PATH\" wezterm-gui start --always-new-process -- " & cmd
```

I think it would be a little more easier if you post the log file here (after WezTerm fails to open). But I don't know it's location on macOS. @wez? Meanwhile, you can redirect stdout and stderr to some file (see below). After WezTerm fails to open, check that file (`/tmp/wezterm-out.txt`) for useful msg.
```applescript
do shell script ">>/tmp/wezterm-out.txt 2>&1 env PATH=/usr/local/bin:\"$PATH\" wezterm-gui start --always-new-process -- " & cmd
```

**Reply by wez** [reply]
The installed path depends on how it was installed; there's no one standard place. It may be `/Applications/WezTerm.app/Contents/MacOS/wezterm` or it may be run out of the extracted zip. homebrew probably puts it in `/usr/local/bin`. The nix folks could put it somewhere like `$HOME/.nix-profile/bin/`.

**Reply by twisteria** [reply]
Thank you all!
Just tried your suggestions and here is the result:

The latest script I got below log is here for your reference:
```
on run {input, parameters}
	set cmd to "nvim"
	if input is not {} then
		set filePath to POSIX path of input
		set cmd to "env PATH=/usr/local/bin:\"$PATH\" nvim \"" & filePath & "\""
	end if
	do shell script ">>/tmp/wezterm-out.txt 2>&1 env PATH=/usr/local/bin:\"$PATH\" wezterm-gui start --always-new-process -- " & cmd
end run
```
and got below log:
```
13:04:18.057  WARN   wezterm_term::terminalstate > unhandled XtermKeyMode OtherKeys Some(2)
13:04:18.143  INFO   wezterm_gui::termwindow     > OpenGL initialized! Intel(R) Iris(TM) Plus Graphics 655 4.1 INTEL-18.7.4 is_context_loss_possible=false wezterm version: 20220624-141144-bd1b7c5d
13:04:23.214  ERROR  wezterm_gui::glyphcache     > load_glyph failed; using blank instead. Error: load_and_render_glyph: FT_Load_Glyph glyph_index:261: FreeType error 36 0x24: invalid size handle. GlyphInfo { is_space: false, num_cells: 1, cluster: 0, font_idx: 5, glyph_pos: 261, x_advance: 0.0, y_advance: 0.0, x_offset: 0.0, y_offset: 0.0 } TextStyle { font: [FontAttributes { family: "UDEV Gothic NF", weight: FontWeight(700), stretch: Normal, style: Normal, is_fallback: false, is_synthetic: false, harfbuzz_features: None, freetype_load_target: None, freetype_render_target: None, freetype_load_flags: None, scale: None }], foreground: None }
13:06:30.075  ERROR  wezterm_gui::glyphcache     > load_glyph failed; using blank instead. Error: load_and_render_glyph: FT_Load_Glyph glyph_index:1039: FreeType error 36 0x24: invalid size handle. GlyphInfo { is_space: false, num_cells: 1, cluster: 0, font_idx: 2, glyph_pos: 1039, x_advance: 0.0, y_advance: 0.0, x_offset: 0.0, y_offset: 0.0 } TextStyle { font: [FontAttributes { family: "UDEV Gothic NF", weight: FontWeight(700), stretch: Normal, style: Normal, is_fallback: false, is_synthetic: false, harfbuzz_features: None, freetype_load_target: None, freetype_render_target: None, freetype_load_flags: None, scale: None }], foreground: None }
13:06:33.632  WARN   wezterm_term::terminalstate > unhandled XtermKeyMode OtherKeys Some(2)
13:06:33.736  INFO   wezterm_gui::termwindow     > OpenGL initialized! Intel(R) Iris(TM) Plus Graphics 655 4.1 INTEL-18.7.4 is_context_loss_possible=false wezterm version: 20220624-141144-bd1b7c5d
13:07:37.545  ERROR  wezterm_gui::glyphcache     > load_glyph failed; using blank instead. Error: load_and_render_glyph: FT_Load_Glyph glyph_index:1039: FreeType error 36 0x24: invalid size handle. GlyphInfo { is_space: false, num_cells: 1, cluster: 0, font_idx: 2, glyph_pos: 1039, x_advance: 0.0, y_advance: 0.0, x_offset: 0.0, y_offset: 0.0 } TextStyle { font: [FontAttributes { family: "UDEV Gothic NF", weight: FontWeight(700), stretch: Normal, style: Normal, is_fallback: false, is_synthetic: false, harfbuzz_features: None, freetype_load_target: None, freetype_render_target: None, freetype_load_flags: None, scale: None }], foreground: None }
```
what I did during the time was -- opened the first file --> waited 1 min --> (tried to) open second file --> waited 1 min --> closed the first file (the second file automatically opened up) --> waited 1 min --> closed the second file.

I have also tried ```nvim-remote``` ( I believe ```mhinz/neovim-remote``` is the one?)
Installation successful, but got ```zsh: command not found: nvr``` error. Not sure if I should deepdive into this..

**Reply by MuhammedZakir** [reply]
That's wierd. I can't help you anymore on this without testing myself. :-(

About `zsh: command not found: nvr`: this is because `nvr` is not in your `PATH`. Get the installation location (see [1]), and put that in your path. For example, if it is `/path/to/dir` and you're using ZSH, then put `export PATH=/path/to/dir:"$PATH"` in `~/.zshenv`.

[1] https://github.com/mhinz/neovim-remote/blob/master/INSTALLATION.md

**Reply by twisteria** [reply]
Thank you!
You have been a great help anyway!

I'll look into neovim-remote in my spare time.

Thanks a lot!

**Comment by htlin222** [comment]
I happen to be looking for a solution for wezterm + neovim, and I saw this discussion thread.

Actually, the solution is very simple. You just need to use `which nvim` and `which wezterm-gui` to find the `PATH` of these two binary files. Therefore, you only need to change the code as follows:

```applescript
on run {input, parameters}
	set cmd to "nvim"
	if input is not {} then
		set filePath to POSIX path of input
		set cmd to "/opt/homebrew/bin/nvim \"" & filePath & "\""
	end if
	do shell script "/opt/homebrew/bin/wezterm-gui start -- " & cmd
end run
```

**Reply by ChristinWhite** [reply]
While portability isn't really a concern here given the context it's still worth being [cautious](https://unix.stackexchange.com/questions/85249/why-not-use-which-what-to-use-then) [about](https://mywiki.wooledge.org/BashFAQ/081) using `which` in scripts.
